docs/plugin_author_guide/compiler_plugins.rst

```diff
@@ -0,0 +1,237 @@
+    answer = res_opt.compute()
+
+The Metagraph optimizer will leave all non-Metagraph tasks unchanged, so it is
+always safe to apply.
+
+
+Visualizing Compilation
+-----------------------
+
+Metagraph placeholder objects have a custom ``visualize()`` method which works
+the same as the standard `Dask visualize() method`_, but with special shapes
+and labels for Metagraph operations.  For example, this DAG:
+
+.. image:: visualize.png
+
+shows translation steps with ellipses, the concrete type of results with
+parallelograms, and algorithms with octagons.
+
+As with optimization, the custom Metagraph visualize method can be used with
+any Dask object by calling it directly:
+
+.. code-block:: python
+
+    import metagraph as mg
+
+    mg.visualize(my_dask_object)
+
+When the DAG contains compilable tasks, they will be highlighted with a single
+red octagon outline:
+
+.. image:: vis_unopt.png
+
+And when the optimizer has compiled and fused tasks, the tasks will be shown
+in a double octagon outline with a label listing the algorithms that were fused:
+
+.. image:: vis_opt.png
+
+By default, the visualizer optimizes the graph before drawing it.  To disable
+this, pass ``optimize_graph=False`` to the ``visualize()`` method.
+
+.. _dask visualize() method: https://docs.dask.org/en/latest/graphviz.html
+
+
+A Compiler Example
+------------------
+
+Consider the following example of task graph:
+
+.. image:: dag_orig.png
+
+The user who constructed this graph will have in their script a Metagraph
+placeholder object which contains a “future” representing the result of
+“Metagraph Op #7”.  When they pass the future to the “Save Result” function,
+which needs to performs I/O, this will call ``dask.compute()`` on the future to
+produce the result, and the following actions will happen:
+
+1. Dask will attempt to optimize the graph by calling ``__dask_optimize__()``
+on the Metagraph placeholder object.
+
+2. This will invoke the Metagraph optimizer, which will scan the graph for
+Metagraph operations that have concrete implementations which have registered
+themselves as JIT-compilable with a compiler backend. In our example, the
+compilable operations are marked in a bold red outline:
+
+.. image:: dag_compile_highlight.png
+
+3. The optimizer will use some heuristic to select connected subgraphs from
+the compilable nodes.  The current approach fuses maximal linear chains, so as
+to not reduce task-level parallelism in the task graph.  These subgraphs are:
+
+.. image:: dag_compile_subgraphs.png
+
+Note that compilable subgraphs must not contain any translation operations, as
+those are not assumed to have JIT-compilable implementations.  (This could
+change in the future.)
+
+4. Each subgraph will be passed to the compiler backend separately, along with
+any captured arguments.  The backend will build a subprogram by flattening the
+subgraph into a linear call sequence and combining the IR from all of the
```

Comment:
Maybe it'd be a good idea to say "The backend will build a subprogram by compiling each operation in the subgraph into the compiler's IR and then flattening the..." since it wasn't clear on my first reading what IR it's referring to or where this IR comes from.

Perhaps linking to https://en.wikipedia.org/wiki/Intermediate_representation would be nice as well (but isn't necessary).

Reply:
I rewrote this paragraph to be more clear about what is happening.